[GeoMechanicsApplication] Couple heat and pressure-deformation elements by CoSim

mnabideltares opened the pull request:
**📝 Description**
Themal element is already implemented, but standalone, without water pressure.
However, this element has to be coupled with the water pressure and deformation.
Water pressure induce convection, and hence the convection term has to be included.

**🆕 Changelog**
- Added wapper for GeoMechanicsApplication in Cosim application
- Added convection term to the thermal element
- Added water pressure effects to the dispersion matrix.
- Modify the input in order to read the required variables
- Add functions for variable density and viscosity, because of thermal effects
- Link the thermal element to the variable density
- Link the pressure/deformation related elements to variable density and viscosity

mnabideltares committed:
geomechanics wrapper is added, Thermal element is modified

mnabideltares committed:
Coupled the thermal effects to pressure, by means of water density and viscosity.

A new keyword is added to pressure and thermal input, "thermal_pressure_coupled":false, need to be true for coupling by cosim.

mnabideltares committed:
Coupled thermal effects to UPwSmallStrainElement

mnabideltares committed:
Added "thermal_pressure_coupled" : false  to UPw solver

mnabideltares committed:
A typo fix

mnabideltares committed:
"thermal_pressure_coupled": false, is added to geomechanics_solver.py

mnabideltares committed:
3 conditions are added: UPwNormalFluxCondition2D3N , UPwNormalFluxCondition2D4N and UPwNormalFluxCondition2D5N

mnabideltares committed:
Fixed to skip discharge calculation, when there is no flow/pressure coupled

mnabideltares committed:
PwNormalFluxCondition2D3N, PwNormalFluxCondition2D4N, PwNormalFluxCondition2D5N are added

mnabideltares committed:
Merge branch 'geo/10372-HeatTransfer-temp' into geo/10372-HeatTransfer-temp-branch-cosim

# Conflicts:
#	applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.cpp
#	applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.hpp

mnabideltares committed:
addressed some of the code smells

mnabideltares committed:
Code duplication in Pw_normal_flux_condition.cpp is replaced by a smaller function

mnabideltares committed:
Making a class ThermalUtilities to avoid code duplications for elements

mnabideltares committed:
Added: UPwNormalFaceLoadCondition2D3N, UPwNormalFaceLoadCondition2D4N, UPwNormalFaceLoadCondition2D5N

mnabideltares committed:
Optimized a part of code in UPwNormalFaceLoadCondition

mnabideltares committed:
More Optimization in UPwNormalFaceLoadCondition

mnabideltares committed:
Merge branch 'geo/10372-HeatTransfer-temp' into geo/10372-HeatTransfer-temp-branch-cosim

# Conflicts:
#	applications/GeoMechanicsApplication/custom_elements/U_Pw_small_strain_element.cpp
#	applications/GeoMechanicsApplication/tests/test_helper.py

mnabideltares committed:
Bug fix after merge

Merging led to wrong results, it was due a missing line (was removed in a previous commit #11471)

philbucher commented on applications/CoSimulationApplication/python_scripts/solver_wrappers/kratos/geomechanics_wrapper.py:
```diff
@@ -0,0 +1,39 @@
+        # now we know that the solver uses MPI, only question left is whether to use all ranks or a subset
+        self._CheckDataCommunicatorIsConsistentlyDefined(self.project_parameters["solver_settings"]["model_import_settings"], self.settings["mpi_settings"])
+
+        return super()._GetDataCommunicator()
```
I think this is not correct/handled better by the baseclass. Have you tested this/had problems?

mnabideltares replied:
Hi @philbucher 
Thank you for your review. It is highly appreciated.
Can you please elaborate in more details what is the issue in this script? For consistency, I tried to built it as similar as possible to the scripts for other applications. 
The code runs and converges. We are trying to validate the results, however this is still on progress.
Do I need to apply changes in this script?

philbucher replied:
hey, this is only about MPI. Are you using the CoSimApp in MPI? This particular part of the code is used to enable solvers to use less ranks than `MPI_COMM_WORLD` has.

I would like to understand what you need differently from e.g. the fluid or structural wrapper

mnabideltares replied:
Hi @philbucher 
Thank you for your explanation.
We are not using CoSimApp with MPI, as the MPI is still not ready in the GeoMechanics application. So, I paid less attention to this part of the script.

However, it is preferly to also fix this part in order to make parts correct and ready, in the case we fix/add the MPI in GeoMechApp in the future. 

It is still not clear to me what we need differently from the other opplications, but as far as I can imagine, we will use the standard MPI, like the most of the other applications.

WPK4FEM commented on applications/CoSimulationApplication/python_scripts/solver_wrappers/kratos/geomechanics_wrapper.py:
```diff
@@ -0,0 +1,39 @@
+    return GeoMechanicsWrapper(settings, model, solver_name)
+
+class GeoMechanicsWrapper(kratos_base_wrapper.KratosBaseWrapper):
+    """This class is the interface to the StructuralMechanicsApplication of Kratos"""
```
StructuralMechanics --> GeoMechanics

WPK4FEM commented on applications/GeoMechanicsApplication/custom_conditions/Pw_condition.cpp:
```diff
@@ -161,5 +161,7 @@ template class PwCondition<3,3>;
 template class PwCondition<3,4>;
 
 template class PwCondition<2,3>;
+template class PwCondition<2,4>;
```
Is the 2,4 ever used ( In other words are the plane elements with 4 nodes per side available?)

WPK4FEM commented on applications/GeoMechanicsApplication/custom_conditions/Pw_normal_flux_condition.cpp:
```diff
@@ -88,66 +88,32 @@ void PwNormalFluxCondition<TDim,TNumNodes>::
 
 template class PwNormalFluxCondition<2,2>;
+template class PwNormalFluxCondition<2,3>;
+template class PwNormalFluxCondition<2,4>;
```
Is 2,4 ever used ( 4 nodes on an edge)?

WPK4FEM commented on applications/GeoMechanicsApplication/custom_conditions/Pw_normal_flux_condition.cpp:
```diff
@@ -88,66 +88,32 @@ void PwNormalFluxCondition<TDim,TNumNodes>::
+template class PwNormalFluxCondition<2,4>;
+template class PwNormalFluxCondition<2,5>;
 template class PwNormalFluxCondition<3,3>;
 template class PwNormalFluxCondition<3,4>;
```
No need for 3,6 and 3,8, a quadratic mesh of solids has 6 node triangular or 8 node quadrilateral faces.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_conditions/T_condition.cpp:
```diff
@@ -0,0 +1,157 @@
+template<unsigned int TDim, unsigned int TNumNodes>
+void TCondition<TDim,TNumNodes>::GetDofList(
+    DofsVectorType& rConditionDofList,
+    const ProcessInfo& rCurrentProcessInfo) const
```
rCurrentProcessInfo is not used here, why is it an argument

WPK4FEM commented on applications/GeoMechanicsApplication/custom_conditions/T_condition.cpp:
```diff
@@ -0,0 +1,157 @@
+    KRATOS_TRY
+
+    unsigned int conditionSize = TNumNodes;
+    const GeometryType& rGeom = GetGeometry();
```
rGeom is first used at line 68. To keep it as short lived as possible, its declaration can be just before the for loop.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_conditions/T_condition.cpp:
```diff
@@ -0,0 +1,157 @@
+    if (rConditionDofList.size() != conditionSize)
+        rConditionDofList.resize( conditionSize );
+
+    for (unsigned int i = 0; i < TNumNodes; ++i) {
```
The natural extent of the loop is conditionSize according to the resize on line 65.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_conditions/T_condition.cpp:
```diff
@@ -0,0 +1,157 @@
+
+    unsigned int conditionSize = TNumNodes;
+    //Resetting the LHS
+    if ( rLeftHandSideMatrix.size1() != conditionSize )
```
This implicitely assumes that if size1 is correct also size2 is correct.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_conditions/T_condition.cpp:
```diff
@@ -0,0 +1,157 @@
+    noalias( rLeftHandSideMatrix ) = ZeroMatrix( conditionSize, conditionSize );
+
+    //Resetting the RHS
+    if ( rRightHandSideVector.size() != conditionSize )
```
For vectors and matrices the pattern we see here:
- check the size and adjust if necessary
- fill with zeros
appears often. This may be factored out.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_conditions/T_condition.cpp:
```diff
@@ -0,0 +1,157 @@
+// ============================================================================================
+// ============================================================================================
+template<unsigned int TDim, unsigned int TNumNodes>
+void TCondition<TDim,TNumNodes>::EquationIdVector(
```
The comments for GetDofList also apply for EquationIdVector.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_conditions/T_condition.cpp:
```diff
@@ -0,0 +1,157 @@
+template class TCondition<3,3>;
+template class TCondition<3,4>;
+
+template class TCondition<2,3>;
```
could be close to line 150 such that all 2D TCondition are together.
I would also expect the 3,6 and 3,8 for quadratic triangular and quadrilateral faces.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_conditions/T_condition.cpp:
```diff
@@ -0,0 +1,157 @@
+    VectorType& rRightHandSideVector,
+    const ProcessInfo& CurrentProcessInfo)
+{
+    this->CalculateRHS(rRightHandSideVector, CurrentProcessInfo);
```
For CalculateAll, I would expect that it contains setup of the LHS matrices and the right hand side.
Why is only RHS sufficient?

WPK4FEM commented on applications/GeoMechanicsApplication/custom_conditions/T_condition.hpp:
```diff
@@ -0,0 +1,111 @@
+//
+//  Main authors:    Mohamed Nabi
+//                   
+//                   
```
More of the // lines only elongates the file.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_conditions/T_condition.hpp:
```diff
@@ -0,0 +1,111 @@
+#pragma once
+
+// System includes
+#include <cmath>
```
Is this necessary?, I can't find its use. Same remark for condition_utilities.hpp and geo_mechanics_application_variables.h

WPK4FEM commented on applications/GeoMechanicsApplication/custom_conditions/T_normal_flux_condition.cpp:
```diff
@@ -0,0 +1,144 @@
+        Variables.NormalFlux = 0.0;
+        for (unsigned int i = 0; i < TNumNodes; ++i) {
+            Variables.NormalFlux += NContainer(GPoint,i) * normal_flux_vector[i];
+        }
```
Inproduct of shapefunction and nodal flux. Should be 1 line call to an inner product.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_conditions/T_normal_flux_condition.cpp:
```diff
@@ -0,0 +1,144 @@
+    for(unsigned int i = 0; i < TNumNodes; ++i)
+    {
+        normal_flux_vector[i] = Geom[i].FastGetSolutionStepValue(NORMAL_HEAT_FLUX);
+    }
```
repeating pattern to get a variable for geometry nodes

WPK4FEM commented on applications/GeoMechanicsApplication/custom_conditions/T_normal_flux_condition.cpp:
```diff
@@ -0,0 +1,144 @@
+// ============================================================================================
+// ============================================================================================
+template<unsigned int TDim, unsigned int TNumNodes>
+void TNormalFluxCondition<TDim,TNumNodes>::CalculateRHS(
```
in this function I am confused about what is the integration on an edge or surface of the element for the flux and the integration of the surface or volume of the element.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_conditions/T_normal_flux_condition.cpp:
```diff
@@ -0,0 +1,144 @@
+        }
+        
+        //Obtain Np
+        noalias(Variables.Np) = row(NContainer,GPoint);
```
what does the p mean here, I think we have temperature D.O.F. here

WPK4FEM commented on applications/GeoMechanicsApplication/custom_conditions/T_normal_flux_condition.cpp:
```diff
@@ -0,0 +1,144 @@
+template class TNormalFluxCondition<2,2>;
+template class TNormalFluxCondition<2,3>;
+template class TNormalFluxCondition<3,3>;
+template class TNormalFluxCondition<3,4>;
```
Quadratic faces on quadratic solids missing?

WPK4FEM commented on applications/GeoMechanicsApplication/custom_conditions/T_normal_flux_condition.cpp:
```diff
@@ -0,0 +1,144 @@
+    noalias(rVariables.TVector) = rVariables.NormalFlux * rVariables.Np * rVariables.IntegrationCoefficient;
+
+    GeoElementUtilities::
+        AssemblePBlockVector<0, TNumNodes>(rRightHandSideVector, rVariables.TVector);
```
Why use AssemblePBlockVector for a T condition. If the vector only has T components I see no separate blocks.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_conditions/T_normal_flux_condition.hpp:
```diff
@@ -0,0 +1,113 @@
+    {
+        double NormalFlux;
+        double IntegrationCoefficient;
+        array_1d<double,TNumNodes> Np;
```
Why Np i.s.o just N or Nt for interpolation polynomial for temperature.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_conditions/U_Pw_normal_face_load_condition.cpp:
```diff
@@ -72,123 +73,57 @@ void UPwNormalFaceLoadCondition<TDim,TNumNodes>::
-
-    for (unsigned int i=0; i<2; ++i) {
-        NormalStress     += NContainer(GPoint,i)*Variables.NormalStressVector[i];
-        TangentialStress += NContainer(GPoint,i)*Variables.TangentialStressVector[i];
+    for (unsigned int i = 0; i < TNumNodes; ++i) {
+        NormalStress += NContainer(GPoint, i) * Variables.NormalStressVector[i];
     }
```
inner product.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_conditions/U_Pw_normal_face_load_condition.cpp:
```diff
@@ -72,123 +73,57 @@ void UPwNormalFaceLoadCondition<TDim,TNumNodes>::
+        double TangentialStress = 0.0;
+        for (unsigned int i = 0; i < TNumNodes; ++i) {
+            TangentialStress += NContainer(GPoint, i) * Variables.TangentialStressVector[i];
+        }
```
if the NormalStress and TangentialStress form a vector i.s.o. 2 loods doubles, computing the rTractionVector should become a vector operation for which it should not matter anymore wheter it is 2D or 3D

WPK4FEM commented on applications/GeoMechanicsApplication/custom_conditions/U_Pw_normal_face_load_condition.cpp:
```diff
@@ -197,9 +132,12 @@ double UPwNormalFaceLoadCondition<TDim,TNumNodes>::
+template class UPwNormalFaceLoadCondition<2,4>;
+template class UPwNormalFaceLoadCondition<2,5>;
 template class UPwNormalFaceLoadCondition<3,3>;
 template class UPwNormalFaceLoadCondition<3,4>;
```
triangular and quadrilateral faces of quadratic solids missing?

WPK4FEM commented on applications/GeoMechanicsApplication/custom_conditions/U_Pw_normal_flux_condition.cpp:
```diff
@@ -91,6 +91,9 @@ void UPwNormalFluxCondition<TDim,TNumNodes>::
+template class UPwNormalFluxCondition<2,4>;
+template class UPwNormalFluxCondition<2,5>;
 template class UPwNormalFluxCondition<3,3>;
 template class UPwNormalFluxCondition<3,4>;
```
faces of quadratic solids?

WPK4FEM commented on applications/GeoMechanicsApplication/custom_constitutive/thermal_dispersion_2D_law.cpp:
```diff
@@ -0,0 +1,59 @@
+        const double thermalCondSolidXX = rProp[THERMAL_CONDUCTIVITY_SOLID_XX];
+        const double thermalCondSolidXY = rProp[THERMAL_CONDUCTIVITY_SOLID_XY];
+        const double thermalCondSolidYX = rProp[THERMAL_CONDUCTIVITY_SOLID_YX];
+        const double thermalCondSolidYY = rProp[THERMAL_CONDUCTIVITY_SOLID_YY];
```
This way we are just giving the rProp[] entry another name. We could form a tensor/matrix from it directly such that then the forming of conduction matrix C reads like the mathematics:
C = cSolid * thermalConductionSolid + cWater * thermalCondWater
( where thermalsCondWater = rProp[THERMAL_CONDUCTIVITY_WATER] * UnitMatrix )

WPK4FEM commented on applications/GeoMechanicsApplication/custom_constitutive/thermal_dispersion_2D_law.hpp:
```diff
@@ -0,0 +1,106 @@
+
+///@}
+///@name  Enum's
+///@{
```
Do we keep these comments for documentation that seem not to be used?

WPK4FEM commented on applications/GeoMechanicsApplication/custom_constitutive/thermal_dispersion_2D_law.hpp:
```diff
@@ -0,0 +1,106 @@
+    ///@}
+    ///@name Serialization
+    ///@{
+    friend class Serializer;
```
I would expect the include of serializer.h for this.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/U_Pw_small_strain_element.cpp:
```diff
@@ -1121,7 +1122,13 @@ void UPwSmallStrainElement<TDim,TNumNodes>::
+        if (rGeom[0].SolutionStepsDataHas(TEMPERATURE))
+        {
+            Variables.FluidDensity = ThermalUtilities::CalculateWaterDensityOnIntegrationPoints<TDim, TNumNodes>(Variables.Np, rGeom);
+            Variables.DynamicViscosityInverse = ThermalUtilities::CalculateWaterViscosityOnIntegrationPoints<TDim, TNumNodes>(Variables.Np, rGeom);
```
This means that CalculateWaterViscosityOnIntegrationPoints does not do what its name says, but 1/viscosity

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/transient_Pw_element.cpp:
```diff
@@ -508,6 +509,13 @@ void TransientPwElement<TDim,TNumNodes>::
+        if (Geom[0].SolutionStepsDataHas(TEMPERATURE))
+        {
+            Variables.FluidDensity = ThermalUtilities::CalculateWaterDensityOnIntegrationPoints<TDim, TNumNodes>(Variables.Np, Geom);
+            Variables.DynamicViscosityInverse = ThermalUtilities::CalculateWaterViscosityOnIntegrationPoints<TDim, TNumNodes>(Variables.Np, Geom);
```
See comment in the transient_UPw_small_strain_element

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/transient_Pw_element.cpp:
```diff
@@ -508,6 +509,13 @@ void TransientPwElement<TDim,TNumNodes>::
+        }
+
         //Contributions to the left hand side
         if (CalculateStiffnessMatrixFlag) this->CalculateAndAddLHS(rLeftHandSideMatrix, Variables);
```
strange name for the flag, it steers computation of the hydraulic condictivity matrix.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/transient_Pw_element.hpp:
```diff
@@ -175,6 +175,7 @@ class KRATOS_API(GEO_MECHANICS_APPLICATION) TransientPwElement :
     void CalculateAndAddCompressibilityFlow(VectorType &rRightHandSideVector, ElementVariables &rVariables) override;
 
     unsigned int GetNumberOfDOF() const override;
```
SizeType seems more natural than unsigned int ( even SizeType is defined as unsigned int.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.cpp:
```diff
@@ -0,0 +1,745 @@
+    {
+        KRATOS_TRY
+
+    	const GeometryType& rGeom = this->GetGeometry();
```
keep declaration close to its use ( e.g. on line 85 )

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.cpp:
```diff
@@ -0,0 +1,745 @@
+    template<unsigned int TDim, unsigned int TNumNodes>
+    TransientThermalElement<TDim, TNumNodes> ::
+    TransientThermalElement(IndexType NewId,
+        const NodesArrayType& ThisNodes) : Element(NewId, ThisNodes) {}
```
do we have some kind of agreement on the layout of these constructors?

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.cpp:
```diff
@@ -0,0 +1,745 @@
+
+        const GeometryType& rGeom = this->GetGeometry();
+        const unsigned int N_DOF = this->GetNumberOfDOF();
+        unsigned int index = 0;
```
keep declarations as close as possible to use.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.cpp:
```diff
@@ -0,0 +1,745 @@
+        if (rGeom[0].SolutionStepsDataHas(WATER_PRESSURE))
+        {
+            mIsPressureCoupled = true;
+        }
```
mIsPressureCoupled = rGeom[0].SolutionStepsDataHas(WATER_PRESSURE);

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.cpp:
```diff
@@ -0,0 +1,745 @@
+                KRATOS_ERROR << "missing variable DT_TEMPERATURE on node " << Geom[i].Id() << std::endl;
+
+            if (!Geom[i].HasDofFor(TEMPERATURE))
+                KRATOS_ERROR << "missing variable TEMPERATURE on node " << Geom[i].Id() << std::endl;
```
The message is identical to the one on lines 162-163 How can a user now distinguish between the 2 failures.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.cpp:
```diff
@@ -0,0 +1,745 @@
+        const unsigned int NumGPoints = Geom.IntegrationPointsNumber(this->GetIntegrationMethod());
+
+        // shape functions
+        (rVariables.NContainer).resize(NumGPoints, TNumNodes, false);
```
are these first ( ) necessary? I think they will be seen as a codesmell

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.cpp:
```diff
@@ -0,0 +1,745 @@
+        this->CalculateCapacityMatrix(rVariables.TMatrix, rVariables);
+
+        //Distribute permeability block matrix into the elemental matrix
+        GeoElementUtilities::AssemblePBlockMatrix<0, TNumNodes>(rLeftHandSideMatrix, rVariables.TMatrix);
```
there is no pressure block in a T only matrix?

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.cpp:
```diff
@@ -0,0 +1,745 @@
+
+        this->CalculateConvectionMatrix(rVariables.TMatrix, rVariables);
+
+        //Distribute compressibility block matrix into the elemental matrix
```
probably convection i.s.o compressibility

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.cpp:
```diff
@@ -0,0 +1,745 @@
+            * rVariables.SolidHeatCapacity;
+        noalias(TMatrix) = (cWater + cSolid) * outer_prod(rVariables.N, rVariables.N)
+            * rVariables.IntegrationCoefficient
+            * rVariables.DtTemperatureCoefficient;
```
Apart from the layout, I find this a very clear way of writing.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.cpp:
```diff
@@ -0,0 +1,745 @@
+
+        array_1d<double, TNumNodes> Temp = prod(rVariables.GradNT, rVariables.DischargeVector);
+        TMatrix = outer_prod(rVariables.N, Temp) * rVariables.IntegrationCoefficient
+            * rVariables.WaterDensity * rVariables.WaterHeatCapacity;
```
I guess Temp means temporary here, but is easily confused with temperature.
As is the mathematical writing of an integral, please put the rVariables.IntegrationCoefficient at the end. I think of the multiplicative sequence as c_p * rho * dV

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.cpp:
```diff
@@ -0,0 +1,745 @@
+
+        this->CalculateCapacityVector(rVariables.TMatrix, rVariables.TVector, rVariables);
+
+        //Distribute permeability block vector into elemental vector
```
permeability?

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.cpp:
```diff
@@ -0,0 +1,745 @@
+            * rVariables.SolidHeatCapacity;
+        noalias(TMatrix) = (cWater + cSolid) * outer_prod(rVariables.N, rVariables.N)
+            * rVariables.IntegrationCoefficient;
+
```
repetition of the capacity matrix, can be solved by calling the routine for the capacitymatrix before the multiplication on line 549

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.cpp:
```diff
@@ -0,0 +1,745 @@
+
+        this->CalculateConductivityVector(rVariables.TDimMatrix, rVariables.TMatrix, rVariables.TVector, rVariables);
+
+        //Distribute permeability block vector into elemental vector
```
permeability?

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.cpp:
```diff
@@ -0,0 +1,745 @@
+
+        noalias(TMatrix) = prod(TDimMatrix, trans(rVariables.GradNT))
+            * rVariables.IntegrationCoefficient;
+
```
repetition of constructing the conductivitymatrix, call the routine for it instead.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.cpp:
```diff
@@ -0,0 +1,745 @@
+        PressureGrad = PressureGrad - gravityVector;
+
+        this->CalculatePermiabilityMatrix(rVariables.PermiabilityMatrix, rVariables);
+        rVariables.DischargeVector = -prod(rVariables.PermiabilityMatrix, PressureGrad);
```
spelling: Permiability --> Permeability

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.cpp:
```diff
@@ -0,0 +1,745 @@
+        array_1d<double, TDim> PressureGrad = prod(trans(rVariables.GradNT), PressureVector);
+
+        //Add gravity*water_density to PressureGrad
+        array_1d<double, TDim> gravityVector = rGeom[0].FastGetSolutionStepValue(VOLUME_ACCELERATION)
```
Naming: rGeom[0].FastGetSolutionStepValue(VOLUME_ACCELERATION) gives the gravityvector. Multiplication with rVariables.WaterDensity makes it a weightvector

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.cpp:
```diff
@@ -0,0 +1,745 @@
+        C(0, 0) = rProp[PERMEABILITY_XX] * rVariables.DynamicViscosityInverse;
+        C(0, 1) = rProp[PERMEABILITY_XY] * rVariables.DynamicViscosityInverse;
+        C(1, 0) = rProp[PERMEABILITY_XY] * rVariables.DynamicViscosityInverse;
+        C(1, 1) = rProp[PERMEABILITY_YY] * rVariables.DynamicViscosityInverse;
```
correct, but crowded. Post multiplication reads easier
        C(0, 0) = rProp[PERMEABILITY_XX]; C(0, 1) = rProp[PERMEABILITY_XY];
        C(1, 0) = rProp[PERMEABILITY_XY]; C(1, 1) = rProp[PERMEABILITY_YY];
        C *= rVariables.DynamicViscosityInverse;

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.cpp:
```diff
@@ -0,0 +1,745 @@
+    template class TransientThermalElement<2,9>;
+    template class TransientThermalElement<3,10>;
+    template class TransientThermalElement<3,20>;
+    template class TransientThermalElement<3,27>;
```
strange grouping, why not all 2D first then all 3D

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.cpp:
```diff
@@ -0,0 +1,745 @@
+            break;
+        default:
+            GI_GAUSS = GeometryData::IntegrationMethod::GI_GAUSS_2;
+            break;
```
this default is likely incorrect for some of the 3D mentioned in the template for elements below

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/transient_thermal_element.hpp:
```diff
@@ -0,0 +1,219 @@
+            Matrix NContainer;
+            GeometryType::ShapeFunctionsGradientsType DN_DXContainer;
+
+            // needed for updated Lagrangian:
```
How can a thermal element do updated Lagrangian, it does not contain displacement D.O.F. so it does not deform.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_python/add_custom_strategies_to_python.cpp:
```diff
@@ -60,6 +61,7 @@ void AddCustomStrategiesToPython(pybind11::module& m)
     using NewmarkQuasistaticPwSchemeType = NewmarkQuasistaticPwScheme< SparseSpaceType, LocalSpaceType >;
     using BackwardEulerQuasistaticUPwSchemeType = BackwardEulerQuasistaticUPwScheme< SparseSpaceType, LocalSpaceType >;
     using BackwardEulerQuasistaticPwSchemeType = BackwardEulerQuasistaticPwScheme< SparseSpaceType, LocalSpaceType >;
+	using BackwardEulerQuasistaticTSchemeType = BackwardEulerQuasistaticTScheme< SparseSpaceType, LocalSpaceType >;
```
I understand that when you have a UPw and a Pw schema, consistency dictates that a T scheme is created. However, I think that for something quasistatic, no time integration is needed at all. If a time integration scheme is needed ( for the reason of nonzero first or second order time derivatives ) the choice is between Newmark and BackwardEuler and that has nothing to do with UPw, Pw or T, but only with the order of the differential equations used.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_strategies/schemes/newmark_quasistatic_T_scheme.hpp:
```diff
@@ -0,0 +1,154 @@
+        //check that variables are correctly allocated
+        for (const auto& rNode : rModelPart.Nodes())
+        {
+            if (rNode.SolutionStepsDataHas(TEMPERATURE) == false)
```
if (!rNode.SolutionStepsDataHas(TEMPERATURE))

WPK4FEM commented on applications/GeoMechanicsApplication/custom_strategies/schemes/newmark_quasistatic_T_scheme.hpp:
```diff
@@ -0,0 +1,154 @@
+                                        - rNode.FastGetSolutionStepValue(TEMPERATURE, 1);
+            const auto &PreviousDtTemperature = rNode.FastGetSolutionStepValue(DT_TEMPERATURE, 1);
+
+            rNode.FastGetSolutionStepValue(DT_TEMPERATURE) =  1.0/(mTheta*mDeltaTime)*(DeltaTemperature - (1.0-mTheta)*mDeltaTime*PreviousDtTemperature);
```
Why not using the DT_TEMPERATURE_COEFFICIENT here

WPK4FEM commented on applications/GeoMechanicsApplication/custom_strategies/schemes/newmark_quasistatic_T_scheme.hpp:
```diff
@@ -0,0 +1,154 @@
+        KRATOS_TRY
+
+    	mDeltaTime = rModelPart.GetProcessInfo()[DELTA_TIME];
+        rModelPart.GetProcessInfo()[DT_TEMPERATURE_COEFFICIENT] = 1.0 / (mTheta * mDeltaTime);
```
I would expect that DT_TEMPERATURE_COEFFICIENT and DT_PRESSURE_COEFFICIENT will be the same ( first order differential equation ) and a member variable of the time integration method

WPK4FEM commented on applications/GeoMechanicsApplication/custom_utilities/thermal_utilities.hpp:
```diff
@@ -0,0 +1,75 @@
+//  License:		 BSD License
+//					 Kratos default license: kratos/license.txt
+//
+//  Main authors:    Mohamed Nabi
```
No geomechanics logo?

WPK4FEM commented on applications/GeoMechanicsApplication/custom_utilities/thermal_utilities.hpp:
```diff
@@ -0,0 +1,75 @@
+        return
+            + 9.998396e+2 + 6.764771e-2 * temp - 8.993699e-3 * std::pow(temp, 2)
+            + 9.143518e-5 * std::pow(temp, 3) - 8.907391e-7 * std::pow(temp, 4)
+            + 5.291959e-9 * std::pow(temp, 5) - 1.359813e-11 * std::pow(temp, 6);
```
Any references to where this formula comes from?

WPK4FEM commented on applications/GeoMechanicsApplication/custom_utilities/thermal_utilities.hpp:
```diff
@@ -0,0 +1,75 @@
+    // ============================================================================================
+    // ============================================================================================
+    template<unsigned int TDim, unsigned int TNumNodes>
+    static double CalculateWaterDensityOnIntegrationPoints(
```
peculiar indentation

WPK4FEM commented on applications/GeoMechanicsApplication/custom_utilities/thermal_utilities.hpp:
```diff
@@ -0,0 +1,75 @@
+	
+	
+
+}; // class StructuralMechanicsElementUtilities.
```
copied comment?

WPK4FEM commented on applications/GeoMechanicsApplication/geo_mechanics_application.cpp:
```diff
@@ -206,26 +207,50 @@ void KratosGeoMechanicsApplication::Register() {
+    KRATOS_REGISTER_ELEMENT("TransientThermalElement3D10N", mTransientThermalElement3D10N)
+    KRATOS_REGISTER_ELEMENT("TransientThermalElement3D20N", mTransientThermalElement3D20N)
+    KRATOS_REGISTER_ELEMENT("TransientThermalElement3D27N", mTransientThermalElement3D27N)
+
```
Updated Lagrangian?
what is the ordering here.

WPK4FEM commented on applications/GeoMechanicsApplication/geo_mechanics_application_constants.h:
```diff
@@ -93,4 +93,9 @@ namespace Kratos
+    enum indexHeat2D : int {
+        INDEX_2D_HEAT_X,
+        INDEX_2D_HEAT_Y
+    };
```
Heat flux vector?

WPK4FEM commented on applications/GeoMechanicsApplication/python_scripts/geomechanics_solver.py:
```diff
@@ -160,17 +161,24 @@ def AddVariables(self):
         ## smoothing variables
-        self._add_smoothing_variables()
+        if self.settings["nodal_smoothing"].GetBool():
+            self._add_smoothing_variables()
```
Yes! only add what is needed. That may even be possible for the dynamic_variables.

mnabideltares replied:
No, for dynamic_variable, it is not possible in the current stage. The initialization of Pressure element is always dependent on pressure-deformation, because of inheritance. 
It is indeed prefered to separate them and make pressure to be independent than the dynamic variables. However,  it will be feasible after reverting the inheritance.

WPK4FEM reviewed (commented):
Dear Mohammed,
It took a while to read through, but I have really read through everything and made many comments. My C++ knowledge is not good enough to comment on those aspects, I hope that Anne or Richard will read through as well.
Regards, Wijtze Pieter

mnabideltares committed:
Changes based on reviewers requests

mnabideltares committed:
Added the missing geometries for conditions in the thermal part

mnabideltares committed:
A bug fix

mnabideltares committed:
Adressed points from review requests

mnabideltares committed:
Added the effect of Pw on thermal dispersion

NB: This was added on June 20, but disapperared on the next commit on Jun 27th. Here brought it back but in a class of GeoThermalDispersion2DLaw

mnabideltares committed:
Coupled thermal effects to more U-Pw elements

NB: Not all U-Pw elements are set yet

mnabideltares committed:
Removed templated ThermalUtilities ,  added thermal to more U-Pw elements

mnabideltares committed:
EXPERIMENTAL: Added convection vector to RHS, Added Dt cooefficient to the capacity vector

mnabideltares committed:
Fixed bugs

mnabideltares committed:
Bug fix

mnabideltares committed:
Geomechanics wrapper in CoSim Application is fixed: The MPI part is removed, as GeoMech does not support MPI yet

mnabideltares committed:
Details of implementation of thermo-pressure is added by means a list, for REVIEW purpose

NB: This file has not to be merged in the master

mnabideltares committed:
Fix in README_CoSim.md

mnabideltares committed:
Fixed a bug related to pressure sign convention in the coupling to thermal element

mnabideltares committed:
A new word is added to Pw, U-Pw, and Thermal solvers, "update_density_viscosity" : true,

This is to let the user to be able to update the material properties (density and viscosity) or keep them constants, in the case of coupled thermal-pressure(-deformation) problems.

mnabideltares committed:
Fixed a bug + cleanup

mnabideltares committed:
cleanup
